spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/WeightedServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,135 @@
+
+/**
+ * A {@link ServiceInstanceListSupplier} implementation that uses weights to expand the
+ * instances provided by delegate.
+ *
+ * @author Zhuozhi Ji
+ */
+public class WeightedServiceInstanceListSupplier extends DelegatingServiceInstanceListSupplier {
+
+	private static final Log LOG = LogFactory.getLog(WeightedServiceInstanceListSupplier.class);
+
+	static final String METADATA_WEIGHT_KEY = "weight";
+
+	static final int DEFAULT_WEIGHT = 1;
+
+	private final WeightFunction weightFunction;
+
+	public WeightedServiceInstanceListSupplier(ServiceInstanceListSupplier delegate) {
+		super(delegate);
+		this.weightFunction = WeightedServiceInstanceListSupplier::metadataWeightFunction;
+	}
+
+	public WeightedServiceInstanceListSupplier(ServiceInstanceListSupplier delegate, WeightFunction weightFunction) {
+		super(delegate);
+		this.weightFunction = weightFunction;
+	}
+
+	@Override
+	public Flux<List<ServiceInstance>> get() {
+		return delegate.get().map(this::expandByWeight);
+	}
+
+	private List<ServiceInstance> expandByWeight(List<ServiceInstance> instances) {
+		if (instances.size() == 0) {
+			return instances;
+		}
+
+		int[] weights = instances.stream().mapToInt(instance -> {
+			try {
+				int weight = weightFunction.apply(instance);
+				if (weight <= 0) {
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(String.format(
+								"The weight of the instance %s should be a positive integer, but it got %d, using %d as default",
+								instance.getInstanceId(), weight, DEFAULT_WEIGHT));
+					}
+					return DEFAULT_WEIGHT;
+				}
+				return weight;
+			}
+			catch (Exception e) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug(String.format(
+							"Exception occurred during apply weight function to instance %s, using %d as default",
+							instance.getInstanceId(), DEFAULT_WEIGHT), e);
+				}
+				return DEFAULT_WEIGHT;
+			}
+		}).toArray();
+
+		int gcd = weights[0];
+		int total = weights[0];
+		for (int i = 1; i < weights.length; i++) {
+			gcd = gcd(gcd, weights[i]);
+			total += weights[i];
+		}
+
+		List<ServiceInstance> newInstances = new ArrayList<>(total / gcd);
+
+		// use iterator for some implementation of the List that not supports
+		// RandomAccess, but `weights` is supported, so use a local variable `i`
+		// to get the current position.
+		int i = 0;
+		for (ServiceInstance instance : instances) {
+			weights[i] /= gcd;
+			for (int weight = 0; weight < weights[i]; weight++) {
+				newInstances.add(instance);
+			}
+			i++;
+		}
+
+		Collections.shuffle(newInstances);
+		return newInstances;
+	}
+
+	static int metadataWeightFunction(ServiceInstance serviceInstance) {
+		Map<String, String> metadata = serviceInstance.getMetadata();
+		if (metadata != null) {
+			String weightValue = metadata.get(METADATA_WEIGHT_KEY);
+			if (weightValue != null) {
+				return Integer.parseInt(weightValue);
+			}
+		}
+		return DEFAULT_WEIGHT;
+	}
+
+	static int gcd(int a, int b) {
```

Comment:
`gcd` is just a function to calculate the Greatest Common Divisor of `a` and `b` to keep list as small as possible to reduce memory usage

Reply:
For the kind of enterprise applications we work with, we usually don't need to save memory usage so drastically as to restrict the variable naming to single letters. Instead, it's very important to maintain easy readability, though we can leave it as  `a` and `b` in this case as we just process 2 numbers. Please rename the method to `greatestCommonDivisor`.

Reply:
I concur with Olga. I would go so far as to rename `w` to `weight`, `a` to int1, and `b` to `int2`. At some point, someone else (and possibly you, years from now) has to figure out what you did. You want that process to be as fast and as smooth as possible. Clear naming is a big help with that, especially in an open source community with hundreds of thousands of members whose abilities vary widely.

Reply:
> For the kind of enterprise applications we work with, we usually don't need to save memory usage so drastically as to restrict the variable naming to single letters. Instead, it's very important to maintain easy readability, though we can leave it as `a` and `b` in this case as we just process 2 numbers. Please rename the method to `greatestCommonDivisor`.

I'm sorry, I realize this might be a problem with my expression. The memory to be saved here is not the size of the source code or the compiled constant pool, but because after we expand the elements according to the weight, there will be many repeated elements, and the weight ratio will not be affected after these repeated elements are deleted. But it will reduce the memory usage of the list, and the ratio of this reduction is the greatest common divisor

Reply:
Thanks, @OlgaMaciaszek and @Buzzardo , I will rename method to `greatestCommonDivisor`. 
But I don't think variable name with type is a good idea (of course, just for me), I will keep the name of arguments, just like `java.lang.Math#max(int a, int b)`, what is also a purely mathematical function.